orion/primitives/azure_anomaly_detector.py

```diff
@@ -0,0 +1,175 @@
+        target_column (int):
+            Indicating which column of X is the target.
+        sequence_size (int):
+            Length of the input sequences.
+        pad_size (int):
+            Length of the values from previous window.
+
+    Returns:
+        list:
+            * A list of sliced value ndarray.
+            * A list of sliced index ndarray.
+    """
+    X_ = list()
+    index_ = list()
+
+    min_pad = 0
+    start = 0
+    max_start = len(X) - 1
+
+    target = X[:, target_column]
+
+    while start < max_start:
+        end = start + sequence_size
+
+        X_.append(target[start - min_pad:end])
+        index_.append(index[start - min_pad:end])
+
+        start = end
+        min_pad = pad_size
+
+    return np.asarray(X_), np.asarray(index_)
+
+
+def detect_anomalies(X, index, interval, pad_size, subscription_key, endpoint, granularity,
+                     custom_interval=None, period=None, max_anomaly_ratio=None, sensitivity=None):
+    """Microsoft's Azure Anomaly Detection tool.
+
+    Args:
+        X (ndarray):
+            N-dimensional array containing the input value sequences.
+        index (ndarray):
+            N-dimensional array containing the input index sequences.
+        interval (int):
+            Integer denoting time span frequency of the data.
+        pad_size (int):
+            Length of the values from previous sequence.
+        subscription_key (str):
+            Resource key for authenticating your requests.
+        endpoint (str):
+            Resource endpoint for sending API requests.
+        granularity (str or Granularity):
+            Can only be one of yearly, monthly, weekly, daily, hourly or minutely.
+            Granularity is used for verify whether input series is valid. Possible values
+            include: 'yearly', 'monthly', 'weekly', 'daily', 'hourly', 'minutely'.
+        custom_interval (int):
+            Integer used to set non-standard time interval, for example, if the series
+            is 5 minutes, request can be set as {"granularity":"minutely", "custom_interval":5}.
+            If not given, `None` is used.
+        period (int):
+            Periodic value of a time series. If not given, `None` is used, and the API
+            will determine the period automatically.
+        max_anomaly_ratio (float):
+            Advanced model parameter, max anomaly ratio in a time series. If not given,
+            `None` is used.
+        sensitivity (int):
+            Advanced model parameter, between 0-99, the lower the value is, the larger
+            the margin value will be which means less anomalies will be accepted. If not given,
+            `None` is used.
+
+        Returns:
+            list:
+                Array containing start-index, end-index, score for each anomalous sequence.
+                Note that the API does not have an anomaly score, and so score is set to `None`.
+    """
+
+    def _convert_date(x):
```

Comment:
This `_convert_date` function is not using any information that is known only during the execution of the `detect_anomalies` one, so it can be defined only once at module level instead of re-defining in on the fly every time `detect_anomalies` is called:

```
def _convert_date(x):
    ...

def detect_anomalies(...):
    ...
```